--- a/usbmount.conf
+++ b/usbmount.conf
@@ -35,7 +35,7 @@ FILESYSTEMS="vfat ext2 ext3 ext4 hfsplus ntfs"
 #############################################################################
 # Mount options: Options passed to the mount command with the -o flag.
 # See the warning above regarding removing "sync" from the options.
-MOUNTOPTIONS="sync,noexec,nodev,noatime,nodiratime"
+MOUNTOPTIONS="noexec,nodev,noatime,nodiratime"
 
 # Filesystem type specific mount options: This variable contains a space
 # separated list of strings, each which the form "-fstype=TYPE,OPTIONS".
-- 
2.17.1

